src/dmd/dinterpret.d

```diff
@@ -4866,6 +4866,66 @@ public:
                 result = interpret(ce, istate);
                 return;
             }
+            else if (fd.ident == Id._d_newThrowable)
+            {
+                // Treat `_d_newThrowable!E(args)` as `new E(args)` and interpret `E.ctor(args)`.
+                auto tc = e.type.toBasetype().isTypeClass();
+                auto cd = tc.sym;
+
+                if (!cd.ctor)
+                {
+                    e.error("`%s` cannot be constructed at compile time, because the constructor has no available source code", e.type.toChars());
+                    result = CTFEExp.cantexp;
+                    return;
+                }
+
+                size_t totalFieldCount = 0;
+                for (ClassDeclaration c = cd; c; c = c.baseClass)
+                    totalFieldCount += c.fields.dim;
+                auto elems = new Expressions(totalFieldCount);
+                size_t fieldsSoFar = totalFieldCount;
+                for (ClassDeclaration c = cd; c; c = c.baseClass)
+                {
+                    fieldsSoFar -= c.fields.dim;
+                    foreach (i, v; c.fields)
+                    {
+                        if (v.inuse)
+                        {
+                            e.error("circular reference to `%s`", v.toPrettyChars());
+                            result = CTFEExp.cantexp;
+                            return;
+                        }
+                        Expression m;
+                        if (v._init)
+                        {
+                            if (v._init.isVoidInitializer())
+                                m = voidInitLiteral(v.type, v).copy();
+                            else
+                                m = v.getConstInitializer(true);
+                        }
+                        else
+                            m = v.type.defaultInitLiteral(e.loc);
+                        if (exceptionOrCant(m))
+                            return;
+                        (*elems)[fieldsSoFar + i] = copyLiteral(m).copy();
+                    }
+                }
+
+                // Hack: store a ClassDeclaration instead of a StructDeclaration.
+                auto se = ctfeEmplaceExp!StructLiteralExp(e.loc, cast(StructDeclaration) cd, elems, e.type);
+                se.origin = se;
+                se.ownedByCtfe = OwnedBy.ctfe;
+                Expression eref = ctfeEmplaceExp!ClassReferenceExp(e.loc, se, e.type);
+
+                UnionExp ue = void;
+                auto err = interpretFunction(&ue, cast(FuncDeclaration) cd.ctor, istate, e.arguments, eref);
+                if (exceptionOrCant(err))
+                    return;
+
+                eref.loc = e.loc;
+                result = eref;
+                return;
+            }
```

Comment:
This block seems to be copied from the class branch of `NewExp` - refactor to that code into a function instead of duplicating it here.

Reply:
Wouldn't it be shorter and more expressive, albeit a bit less efficient, if I just create a `NewExp` and interpret it directly?